pyanaconda/dbus/__init__.py

```diff
@@ -1,21 +1,86 @@
+#
+# Representation of DBus connection.
+#
+# Copyright (C) 2017  Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
 import os
 import pydbus
 
 from pyanaconda.anaconda_loggers import get_module_logger
 log = get_module_logger(__name__)
 
-def get_bus():
-    """Get a DBUS bus connection corresponding to the current environment.
+__all__ = ["DBus"]
+
 
-    During normal usage this function should return connection to the system bus,
-    but during testing/development a custom bus might be used.
-    So just always connect to the bus specified by the DBUS_STARTER_ADDRESS
-    environmental variable.
+class DBus(object):
+    """Representation of DBus connection.
+
+    Call DBus.get_connection to get a connection to the DBus, or
+    DBus.publish to publish objects on the DBus, or DBus.get_proxy
+    to get a proxy of a remote DBus object.
     """
-    bus_address = os.environ.get("DBUS_STARTER_ADDRESS")
-    if bus_address:
-        return pydbus.connect(os.environ["DBUS_STARTER_ADDRESS"])
-    else:
-        log.info("DBUS_STARTER_ADDRESS not defined, using system bus")
+    _connection = None
+
+    @staticmethod
+    def get_connection():
+        """Returns a DBus connection."""
+        if not DBus._connection:
```

Comment:
I wonder if there is a chance for a race condition ? Eq. two requests getting two different connections, with the last one being cached.

Reply:
I don't think that could be a problem. The cached connection is just recommended to use, but it is not required.